File: Solutions/CPP/Rotate matrix elements clockwise.cpp
class Solution {
  public:
   vector<vector<int>> rotateMatrix(int M, int N, vector<vector<int>> mat) {
        int row = 0;
        int col = 0;
        int maxRow = M - 1;
        int maxCol = N - 1;
        while(row < maxRow + 1 && col < maxCol + 1){
            if(row == maxRow || col == maxCol){
                break;
            }
            
            int previous = mat[row + 1][col];
            for(int i = col; i <= maxCol; i++){
                int curr = mat[row][i];
                mat[row][i] = previous;
                previous = curr;
            }
            row++;
 
            for(int i = row; i <= maxRow; i++){
                int curr = mat[i][maxCol];
                mat[i][maxCol] = previous;
                previous = curr;
            }
            maxCol--;
        
            if(row < maxRow + 1){
                for(int i = maxCol; i >= col; i--){
                    int curr = mat[maxRow][i];
                    mat[maxRow][i] = previous;
                    previous = curr;
                }
                maxRow--;
            }
         
            if(col < maxCol + 1){
                for(int i = maxRow; i >= row; i--){
                    int curr = mat[i][col];
                    mat[i][col] = previous;
                    previous = curr;
                }
                col++;
            }
        }
        return mat;
    }
};

File: Solutions/Java/Rotate-matrix-elements-clockwise.java
//problem link: https://practice.geeksforgeeks.org/problems/rotate-matrix-elements-clockwise2336/1
class Solution {
  int[][] rotateMatrix(int M, int N, int Mat[][]) {
      // code here
      int row = 0, col = 0;
      int prev, curr;
      while (row < M && col < N)
      {
   
          if (row + 1 == M || col + 1 == N)
              break;
   
          
          prev = Mat[row + 1][col];
   
          
          for (int i = col; i < N; i++)
          {
              curr = Mat[row][i];
              Mat[row][i] = prev;
              prev = curr;
          }
          row++;
   
          
          for (int i = row; i < M; i++)
          {
              curr = Mat[i][N-1];
              Mat[i][N-1] = prev;
              prev = curr;
          }
          N--;
   
          
          if (row < M)
          {  
              for (int i = N-1; i >= col; i--)
              {
                  curr = Mat[M-1][i];
                  Mat[M-1][i] = prev;
                  prev = curr;
              }
          }
          M--;
   
          
          if (col < N)
          {  
              for (int i = M-1; i >= row; i--)
              {
                  curr = Mat[i][col];
                  Mat[i][col] = prev;
                  prev = curr;
              }
          }
          col++;
      }
      return Mat; 
  }
  
}
